origin-growth/src/resources/authentication.js

```diff
@@ -0,0 +1,126 @@
+const _growthModels = require('../models')
+const _identityModels = require('origin-identity/src/models')
+const db = { ..._growthModels, ..._identityModels }
+const logger = require('../logger')
+const Web3 = require('web3')
+const enums = require('../enums')
+const crypto = require('crypto')
+
+const web3 = new Web3(process.env.PROVIDER_URL || 'http://localhost:8545')
+// TODO: have this stored somewhere in the db
+const currentAgreementMessage =
+  'I accept the terms of growth campaign version: 1.0'
+
+/**
+ * Authenticates user's enrollment to the growth campaign
+ * @param {string} referrer - Eth address of the referrer.
+ * @param {Array<string>>} recipients - List of email addresses.
+ */
+async function authenticateEnrollment(accountId, agreementMessage, signature) {
+  if (currentAgreementMessage !== agreementMessage) {
+    throw new Error(
+      `Incorrect agreementMessage. Expected: "${currentAgreementMessage}" received: "${agreementMessage}"`
+    )
+  }
+  const recoveredAccountId = web3.eth.accounts.recover(
+    agreementMessage,
+    signature
+  )
+
+  if (accountId !== recoveredAccountId) {
+    throw new Error('Recovered and provided accounts do not match')
+  }
+
+  const participant = await db.GrowthParticipant.findOne({
+    where: {
+      ethAddress: accountId
+    }
+  })
+
+  if (
+    participant !== null &&
+    participant.status === enums.GrowthParticipantStatuses.Banned
+  ) {
+    logger.warn(`Banned user: ${accountId} tried to enroll`)
+    throw new Error('This user is banned')
+  }
+
+  const authToken =
+    participant === null
+      ? crypto.randomBytes(64).toString('hex')
+      : participant.authToken
+
+  const participantData = {
+    ethAddress: accountId,
+    status: enums.GrowthParticipantStatuses.Active,
+    agreementId: agreementMessage,
+    // if user uses growth from 2 devices let them share the same auth token
```

Comment:
For my understanding, the user will have to agree to the terms on each device, right ? might be worth clarifying in the comment.

Reply:
good point adding a comment